Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/AbilData.xml

```diff
@@ -11022,4 +11022,91 @@
         <AutoCastFilters value="Ground,Massive;Structure,Missile,Destructible,Stasis,Dead,Hidden,Hallucination"/>
         <Arc value="360"/>
     </CAbilEffectTarget>
+    <CAbilMerge id="AP_TyrannozorMerge">
+        <EditorCategories value="Race:Zerg,AbilityorEffectType:Units"/>
+        <CmdButtonArray index="SelectedUnits" DefaultButtonFace="AP_TyrannozorMerge" State="Restricted" Requirements="AP_HaveUltraliskCavern">
+            <Flags index="ToSelection" value="1"/>
+        </CmdButtonArray>
+        <CmdButtonArray index="WithTarget" DefaultButtonFace="AP_TyrannozorMerge" State="Restricted" Requirements="AP_HaveUltraliskCavern"/>
+        <Flags index="IgnoreUnitCost" value="1"/>
+        <Info Unit="AP_Tyrannozor" Time="8"/>
+    </CAbilMerge>
+    <CAbilMerge id="AP_MorphlingToTyrannozorMerge">
+        <EditorCategories value="Race:Zerg,AbilityorEffectType:Units"/>
+        <CmdButtonArray index="SelectedUnits" DefaultButtonFace="AP_MorphlingToTyrannozorMerge" State="Restricted" Requirements="AP_HaveUltraliskCavern">
+            <Flags index="ToSelection" value="1"/>
+        </CmdButtonArray>
+        <CmdButtonArray index="WithTarget" DefaultButtonFace="AP_MorphlingToTyrannozorMerge" State="Restricted" Requirements="AP_HaveUltraliskCavern"/>
+        <Flags index="IgnoreUnitCost" value="0"/>
+        <Info Unit="AP_Tyrannozor" Time="78">
+            <Resource index="Minerals" value="-100"/>

```

Comment:
Should the value be negative here? I'm pretty sure that means it will _return_ 100 minerals rather than cost it.

Is this related to merge ability jank, or can it be adjusted by changing the cost of the Tyrannozor in unit data?

Reply:
Merge uses unit cost unless manually specified.  Since resource efficiency would affect cost, I had the resource efficiency adjust the unit cost and set the morphling merge to -100 to offset the 2x 50 mineral morphling cost.  Seems to work correctly with both resource efficiency and without it.

Reply:
Hm I think I follow. May be worth putting a comment here saying where -100 comes from.

I know @EnvyDragon was fighting resource costs on merge abilities a while ago, I wonder if this would have fixed his issues.

Reply:
Maybe, I was going to use the same method He used for igniters, but it didn't work for me, so I tested some other options.  There isn't a resource efficiency for roachs, so the same issue wasn't present for igniters.

Reply:
I do think that having the Morphling version specifically track Tyrannozor cost, then manually subtracting 100 from that, is a good way to do it if there's a resource efficiency upgrade; that wasn't a problem with roach/igniter, but both Ultras and Tyrannos being impacted by Ultra resource efficiency means that could get complicated pretty quick if you didn't let the game automate it!

Important note: I was specifically fighting _supply_ costs on merge abilities, because the merge effect doesn't check for supply availability at any point in the process; my solution was literally just to increase the morphling's supply cost from 1 to 2, but I'm not thrilled with that choice either way... and if the Tyrannozor uses a whopping _nine food_ then the morphling merge is going to put you over your supply cap pretty quickly. A way to test would be to get like six of them, move them somewhat far apart, cap yourself on supply, and then merge them all. Not sure what the solution would be in this case...

Reply:
Turns out, it doesn't care about the food cost at all, even at over max cap, just keeps going over, which is a problem.  I'm looking for a solution as well.  Funny enough, the tyrannosaur merge actually frees up food, it's a 25% discount over two ultralisks, not sure if that's true for other primal merges or not.  morphlings are really the issue with this, and just increasing the food to 5/6 isn't really a solution (I hope).

Reply:
Primal Igniters do it too (2x2 food roaches -> 3 food igniter) so it's not a huge issue in my opinion; refunding food just means that if the target unit is the one you wanted in the first place, you need a little more supply open to support the intermediary units. It's jank but not terrible.

If you want, you _could_ just disable Tyrannozor morphling merges for now; I have a big dumb workaround in mind involving a button that lets you increase a morphling's food cost (something like a spider mine magazine but the spider mines cost 1 food each, and they can only merge if a validator sees that their food cartridge is full enough?) but I have no idea if it'll work. Either way, it's a complex problem that you _technically_ don't need to solve right now if you don't want to!

Reply:
Could you just make a validator that checks for available food? I think that's something you can do with a validator.

(this is more for Envy, sorry for derail on the PR)

Reply:
I'm looking into that, and I have the validator (I think), but I don't know how to apply it to the ability, cabilmerge doesn't accept straight validators that I see.